[TSK-910] [TSK-911] [TSK-912] NotificationGrid

ObitoDevv opened the pull request:
## Description

<!-- Provide a clear and concise description of the changes made in this pull request for UI-Kit components. -->

- Component name: `NotificationGrid`
- Figma link or screenshot: [Figma Design](https://www.figma.com/design/8KEwRuOoD5IgxTtFAtLlyS/Just_Do_Ad-1.2?node-id=8446-188518&t=Z9yOThOzKCxx7IT2-4)

---


## React Best Practices Checklist

- [x] I have avoided **unnecessary re-renders** by using composition (`children` as props), whenever possible. This applies, in particular, to composite components (like lists of components, and the like).
- [x] **I have designed my components as server-side whenever possible**—I have not used `useState`, `useEffect`, `useContext`, or any other React hook that works with "use client" without a good reason.
- [x] **I have not put props in state**—in general, props should remain as props unless the component needs an `initialValue`, and judiciously manage its own independent state detached from the parent component.
    - If I have stored props in state, I have explained where exactly and justified it with a solid reason here below:
    
- [x] **I have avoided using Context** in any critical component in terms of performance
    - If I have used Context, I have explained where exactly and justified it with a solid reason here below:


## General Checklist

- [x] I have run the lint, build, and test commands locally, as explained in the README, before creating this PR to ensure the code is clean and functional.
- [x] My changes align with the **core models** that have been discussed and agreed upon.
- [x] The component design has been compared to the **Figma design** to ensure consistency.
- [x] The **component is fully responsive** and behaves as expected across different screen sizes.
- [x] The component **supports keyboard web accessibility** (if applicable).
- [x] I have written **unit tests** or **integration tests** to cover the component's functionality above the coverage threshold.
- [x] I have updated the **documentation** for the component, including usage examples and guidelines.
- [x] I have **commented** my code, particularly in complex or non-obvious areas.
- [x] The **component is properly exported** for use in other parts of the application.




---

## Related Issues

<!-- Link any related issue or task. For example: -->
Fixes #134  (issue)

alebg commented on packages/ui-kit/lib/components/grids/notification-grid.tsx:
```diff
@@ -70,31 +123,157 @@ export const NotificationGrid = ({ notifications, onNotificationClick }: Notific
+
+    // Client-side filtering logic for search
+    const doesExternalFilterPass = useCallback(
+        (node: any) => {
+            if (!node.data) return false;
+            const notification = node.data;
```
@ObitoDevv in general please don't use `any` when typing, nor the `as` whenever possible. Please either check how to type this properly, or give us a reason why this typing is not strict

alebg commented on packages/ui-kit/lib/components/grids/notification-grid.tsx:
```diff
@@ -70,31 +123,157 @@ export const NotificationGrid = ({ notifications, onNotificationClick }: Notific
+                    className="flex bg-base-neutral-800 rounded-medium gap-2 text-sm whitespace-nowrap min-w-max"
+                    variant="small"
+                >
+                    <TabTrigger value="all">
```
@ObitoDevv please don't hardcode anything in the ui kit, including platform names. If you need this information, pass it as props so the page can decide what to put as tab values

alebg replied:
@ObitoDevv this needs to variants as explained in the issue. in the CMS variant, we have the tabs with the names of the platforms. In the platform variant, we don't have the tabs so we just render all notifications passed to the grid.

alebg commented on packages/ui-kit/lib/components/grids/notification-grid.tsx:
```diff
@@ -70,31 +123,157 @@ export const NotificationGrid = ({ notifications, onNotificationClick }: Notific
+                        { 'Just Do Ad'} ({notificationCounts.justdoad})
+                    </TabTrigger>
+                    <TabTrigger value="bewerbeag">
+                        { 'Bewerbeag'} ({notificationCounts.bewerbeag})
```
@ObitoDevv Please replace all "bewerbeag" for the correct name, which is "bewerbeagentur"

alebg commented on packages/ui-kit/lib/components/grids/notification-grid.tsx:
```diff
@@ -1,67 +1,120 @@
-ModuleRegistry.registerModules([AllCommunityModule]);
+// Extend TNotification to include platform
+interface ExtendedNotification extends notification.TNotification {
+    platform?: string; // e.g., "justdoad", "bewerbeag", "cms"
```
@ObitoDevv @anilrise11 could you please avoid using so many optionals everywhere, in general? it gets hard to reason about (and hence review) if you do this. In this case I don't see why this would be optional, neither why would `notifications` below be optional.

Note that in the  platform (non-cms) variant of the grid, the ui kit doesn't need to know in which platform it is at all. It will just render all notifications passed to it

alebg reviewed (changes requested):
@ObitoDevv please address my inline comments

ObitoDevv commented:
Hi @alebg sir,

I’ve addressed all feedback. Here’s a brief summary of the changes:

1. Type Safety: Updated doesExternalFilterPass to use `RowNode<ExtendedNotification>` instead of any and typed `node.data` as `ExtendedNotification`.
2. Platform Names: Configured platform names to be passed via the platforms `prop`, removing hardcoded values.
3. Naming Correction: Replaced "`bewerbeag`" with "`bewerbeagentur`" across the component.
4. Variants: Implemented CMS variant with platform tabs and platform variant `(student/coach)` without tabs, rendering all notifications as passed.
5. Non-Optional Types: Ensured platform in `ExtendedNotification `and notifications in `NotificationGridProps `are non-optional for better type safety.
6. Platform Variant Logic: Ensured the platform variant renders all notifications without platform-specific filtering.

All changes are complete, and the component is ready for review. Please let me know if there’s anything else you’d like me to address!

alebg commented:
@ObitoDevv I also don't understand here if this PR is meant to be complete. Why do you have, for example, a `fetchNotificationsWithPlatform` mock function doing an important piece of logic in the official code? 

You don't need to filter anything with pure typescript before passing data to the grid. Instead, you can use the grid itself to filter by platform when you click on the tabs, on the cms variant.

ObitoDevv commented:
Hi @alebg sir, thank you for your feedback and for taking the time to review this PR!
I’d like to clarify a couple of points regarding the concerns you raised:

1. PR Completeness: This PR is not yet complete. My understanding was that the `usergrid PR` would be reviewed and merged first, and I planned to align this PR with the same structure and approach once that was finalized. Apologies for any confusion here—let me know if I should prioritize completing this PR independently or wait for the usergrid PR to be merged.

2. Filtering Logic in fetchNotificationsWithPlatform: Thanks for pointing out that the grid’s built-in filtering can handle platform-specific filtering via the CMS variant tabs. I have a few doubts about how the data is structured and how we should proceed:

- Data Source: Is the data coming from the backend already filtered by platform (e.g., separate API responses for each platform), or is it a mixed dataset containing a platform field? If it’s the latter, I assumed we’d need client-side filtering similar to what we did in the usergrid PR, where we filter users by role in the Tabs.Root component (see below for reference).
- If the backend sends pre-filtered data, I can remove the fetchNotificationsWithPlatform mock and related TypeScript filtering logic. Could you confirm the expected data structure?
- If the data is mixed, I was planning to leverage the grid’s tab-based filtering as shown in the usergrid code below, where we filter users `client-side` based on the selected tab (e.g., student, coach). Would this approach work for the `notifications grid`, or is there a different mechanism you recommend for the` CMS variant`?

```
<Tabs.Root defaultTab="all" onValueChange={handleTabChange}>
    <TabList className="flex bg-base-neutral-800 rounded-medium gap-2 text-sm whitespace-nowrap" variant='small'>
        <TabTrigger value="all" icon={<IconAll />} className="cursor-pointer">
            {dictionary.components.userGrid.all} ({userCounts.all})
        </TabTrigger>
        <TabTrigger value="student" icon={<RoleIcon role="student" />} className="cursor-pointer">
            {dictionary.components.userGrid.students} ({userCounts.student})
        </TabTrigger>
        ...
    </TabList>
    <div className="mt-4">
        <TabContent value="all" className="overflow-auto max-h-[70vh]">
            {renderGridWithActions(props.users || [])}
        </TabContent>
        <TabContent value="student" className="overflow-auto max-h-[70vh]">
            {renderGridWithActions(props.users?.filter(user => user.roles?.some(roleObj => roleObj.role === 'student')) || [])}
        </TabContent>
        ...
    </div>
</Tabs.Root>

```


I’ll work on removing the mock function and refactoring the logic once I have clarity on the data structure and your recommended approach. Could you also advise if there’s anything specific from the usergrid PR that I should replicate here to ensure consistency?

alebg reviewed (commented):
@MytsV I've done a cleanup of this PR, modified the following points:

- Show expand chevron only on truncated messages; works also when resizing the window
- Render filter buttons only on CMS variant
- Allow vertical auto-sizing, instead of setting an amount of rows
- Pass Mark All As Read callback from the outside
- Add 'New' column header for read/unread column, so users can sort on click
- Fix modal date filter
- Show notifications sorted by newer first on initial render
- Remove "label" default from TextInput
- Slightly increase width of filter modal to show all "Bewerbeagentur" string

It should be in line with the UserGrid that we merged already, but it will be good if you take a quick look.

MytsV commented:
@alebg all the little fixes and addressing the flickering is finished
